{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %pip install llama-index-postprocessor-cohere-rerank\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set COHERE API key\n",
    "# import os\n",
    "# os.environ['COHERE_API_KEY'] = '...'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "\n",
    "from llama_index.core.query_pipeline import QueryPipeline\n",
    "from llama_index.core import PromptTemplate\n",
    "from llama_index.postprocessor.cohere_rerank import CohereRerank\n",
    "from llama_index.core.response_synthesizers import TreeSummarize\n",
    "\n",
    "from llama_index.core import Settings\n",
    "from llama_index.llms.azure_openai import AzureOpenAI\n",
    "from llama_index.embeddings.azure_openai import AzureOpenAIEmbedding\n",
    "\n",
    "# You need to deploy your own embedding model as well as your own chat completion model\n",
    "llm = AzureOpenAI(\n",
    "    deployment_name='gpt-35-turbo16k',\n",
    "    model='gpt-35-turbo',\n",
    "    api_key=os.environ['AZURE_OPENAI_API_KEY'],\n",
    "    azure_endpoint=os.environ['AZURE_OPENAI_ENDPOINT'],\n",
    "    api_version=os.environ['AZURE_OPENAI_API_VERSION'],\n",
    ")\n",
    "\n",
    "# llm = AzureOpenAI(\n",
    "#     deployment_name='gpt-4o',\n",
    "#     model='gpt-4o',\n",
    "#     api_key=os.environ['AZURE_OPENAI_API_KEY'],\n",
    "#     azure_endpoint=os.environ['AZURE_OPENAI_ENDPOINT'],\n",
    "#     api_version=os.environ['AZURE_OPENAI_API_VERSION'],\n",
    "# )\n",
    "\n",
    "\n",
    "embed_model = AzureOpenAIEmbedding(\n",
    "    model=\"text-embedding-ada-002\",\n",
    "    deployment_name=\"text-embedding-ada-002\",\n",
    "    api_key=os.environ['AZURE_OPENAI_API_KEY'],\n",
    "    azure_endpoint=os.environ['AZURE_OPENAI_ENDPOINT'],\n",
    "    api_version=os.environ['AZURE_OPENAI_API_VERSION'],\n",
    ")\n",
    "\n",
    "# global settings\n",
    "Settings.llm = llm\n",
    "Settings.embed_model = embed_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datasets import load_dataset\n",
    "import pandas as pd\n",
    "from llama_index.core import Document\n",
    "\n",
    "\n",
    "ds = load_dataset(\"rag-datasets/rag-mini-bioasq\", \"text-corpus\")\n",
    "ds = ds['passages'].to_pandas().set_index('id', drop=True)\n",
    "query_set = load_dataset(\"rag-datasets/rag-mini-bioasq\", \"question-answer-passages\")\n",
    "queries = query_set['test'].take(5)\n",
    "\n",
    "# create a subset of the documents for faster testing\n",
    "\n",
    "passages_required = set()\n",
    "[ passages_required.update([int(id) for id in ids[1:-1].split(\", \")]) for ids in query_set['test'].take(15)['relevant_passage_ids'] ];\n",
    "\n",
    "\n",
    "docs = [Document(text=ds.loc[id].passage, metadata = {'id' : id}) for id in passages_required]\n",
    "for x in docs:\n",
    "    x.doc_id = str(x.metadata['id'])\n",
    "    x.excluded_llm_metadata_keys = ['id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "required_keys={'query_str', 'nodes'} optional_keys=set()\n"
     ]
    }
   ],
   "source": [
    "from llama_index.core import VectorStoreIndex\n",
    " \n",
    "index = VectorStoreIndex.from_documents(docs)\n",
    "\n",
    "def prepare_pipeline(new_index, k, only_retriever=False):\n",
    "    \n",
    "    k = 2*k # k=10 is a good choice as seen in querying notebook, we take double that, rerank them and take top k\n",
    "\n",
    "    # converting vector store to retriever for RAG pipeline, response synthesis will be handled by the summarizer node\n",
    "    retriever = new_index.as_retriever(similarity_top_k=k)  \n",
    "\n",
    "    prompt_str = '''\n",
    "    Expand or rephrase the given query for querying a RAG index, if the query is already quite detailed return the input verbatim.  \n",
    "    QUERY : {query}\n",
    "    '''\n",
    "\n",
    "    prompt_tmpl = PromptTemplate(prompt_str)\n",
    "\n",
    "    reranker = CohereRerank(top_n=k/2)\n",
    "    summarizer = TreeSummarize()\n",
    "    p = QueryPipeline(verbose=True)\n",
    "\n",
    "    # Adding the modules to the pipeline\n",
    "    p.add_modules(\n",
    "        {\n",
    "            \"llm\": llm,\n",
    "            \"prompt_tmpl\": prompt_tmpl,\n",
    "            \"retriever\": retriever,\n",
    "            \"reranker\": reranker,\n",
    "        }\n",
    "    )\n",
    "\n",
    "    if not only_retriever:\n",
    "        p.add_modules({\"summarizer\": summarizer})\n",
    "\n",
    "    #Define the links between modules\n",
    "    p.add_link(\"prompt_tmpl\", \"llm\")\n",
    "    p.add_link(\"llm\", \"retriever\")\n",
    "    p.add_link(\"retriever\", \"reranker\", dest_key=\"nodes\")\n",
    "    p.add_link(\"llm\", \"reranker\", dest_key=\"query_str\")\n",
    "    if not only_retriever:\n",
    "        p.add_link(\"reranker\", \"summarizer\", dest_key=\"nodes\")\n",
    "        p.add_link(\"llm\", \"summarizer\", dest_key=\"query_str\")\n",
    "\n",
    "        # look at summarizer input keys\n",
    "        print(summarizer.as_query_component().input_keys)\n",
    "\n",
    "\n",
    "    return p\n",
    "\n",
    "p = prepare_pipeline(new_index = index, k=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;3;38;2;155;135;227m> Running module prompt_tmpl with input: \n",
      "query: Is Hirschsprung disease a mendelian or a multifactorial disorder?\n",
      "\n",
      "\u001b[0m\u001b[1;3;38;2;155;135;227m> Running module llm with input: \n",
      "messages: \n",
      "    Expand or rephrase the given query for querying a RAG index, if the query is already quite detailed return the input verbatim.  \n",
      "    QUERY : Is Hirschsprung disease a mendelian or a multifactoria...\n",
      "\n",
      "\u001b[0m\u001b[1;3;38;2;155;135;227m> Running module retriever with input: \n",
      "input: assistant: Can you provide information on whether Hirschsprung disease is classified as a mendelian disorder or a multifactorial disorder?\n",
      "\n",
      "\u001b[0m\u001b[1;3;38;2;155;135;227m> Running module reranker with input: \n",
      "query_str: assistant: Can you provide information on whether Hirschsprung disease is classified as a mendelian disorder or a multifactorial disorder?\n",
      "nodes: [NodeWithScore(node=TextNode(id_='972a2f43-ddca-4340-ad71-1b5c00ed6e37', embedding=None, metadata={'id': 23001136}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=['id'], relationships={<...\n",
      "\n",
      "\u001b[0m\u001b[1;3;38;2;155;135;227m> Running module summarizer with input: \n",
      "query_str: assistant: Can you provide information on whether Hirschsprung disease is classified as a mendelian disorder or a multifactorial disorder?\n",
      "nodes: [NodeWithScore(node=TextNode(id_='972a2f43-ddca-4340-ad71-1b5c00ed6e37', embedding=None, metadata={'id': 23001136}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=['id'], relationships={<...\n",
      "\n",
      "\u001b[0mHirschsprung disease is classified as a complex oligo-/polygenic disorder, which suggests that it is a multifactorial disorder rather than a Mendelian disorder. > Source (Doc id: 972a2f43-ddca-4340-ad71-1b5c00ed6e37): Hirschsprung's disease (HSCR) is a fairly frequent cause of intestinal \n",
      "obstruction in children. It is characterized as a sex-linked heterogonous \n",
      "disorder with variable severity and incomplete penetrance giving rise to a \n",
      "variable pattern of inheritance. Although Hirschsprung's disease occurs as an \n",
      "isolated phenotype in at least 70% of cases, it is not infrequently associated \n",
      "with a number of congenital abnormalities and associated syndromes, \n",
      "demonstrating a spectrum of congenital anomali...\n",
      "\n",
      "> Source (Doc id: 1ca2b903-27e8-4e4f-bd1c-c49ccc5ec116): Hirschsprung disease (HSCR), or congenital intestinal aganglionosis, is a common \n",
      "hereditary disorder causing intestinal obstruction, thereby showing considerable \n",
      "phenotypic variation in conjunction with complex inheritance. Moreover, \n",
      "phenotypic assessment of the disease has been complicated since a subset of the \n",
      "observed mutations is also associated with several additional syndromic \n",
      "anomalies. Coding sequence mutations in e.g. RET, GDNF, EDNRB, EDN3, and SOX10 \n",
      "lead to long-segment (L-HS...\n",
      "\n",
      "> Source (Doc id: 110ef15a-726a-44cd-afc7-b8f845dd1c82): Hirschsprung's disease is characterized by the absence of ganglion cells in the \n",
      "myenteric and submucosal plexuses of the gastrointestinal tract. Genetic \n",
      "dissection was successful as nine genes and four loci for Hirschsprung's disease \n",
      "susceptibility were identified. Different approaches were used to find these \n",
      "loci such as classical linkage in large families, identity by descent mapping in \n",
      "an inbred kindred, candidate gene approaches based on naturally occurring mutant \n",
      "mice models, and f...\n",
      "\n",
      "> Source (Doc id: e0ce075b-b922-4b28-b1ab-853b0f13f58f): The major gene for Hirschsprung disease (HSCR) encodes the receptor tyrosine \n",
      "kinase RET. In a study of 690 European- and 192 Chinese-descent probands and \n",
      "their parents or controls, we demonstrate the ubiquity of a >4-fold \n",
      "susceptibility from a C-->T allele (rs2435357: p = 3.9 x 10(-43) in European \n",
      "ancestry; p = 1.1 x 10(-21) in Chinese samples) that probably arose once within \n",
      "the intronic RET enhancer MCS+9.7. With in vitro assays, we now show that the T \n",
      "variant disrupts a SOX10 binding...\n",
      "\n",
      "> Source (Doc id: 2360eafe-1e46-4cc1-8771-2ae4ce088434): Hirschsprung's disease (HSCR) is a congenital disorder of the enteric nervous \n",
      "system and is characterized by an absence of enteric ganglion cells in terminal \n",
      "regions of the gut during development. Dishevelled (DVL) protein is a \n",
      "cytoplasmic protein which plays pivotal roles in the embryonic development. In \n",
      "this study, we explore the cause of HSCR by studying the expression of DVL-1 and \n",
      "DVL-3 genes and their proteins in the aganglionic segment and the ganglionic \n",
      "segment of colon in HSCR p...\n",
      "\n",
      "> Source (Doc id: 7aa70398-da0c-460b-ab42-061f486743e8): Cerebral Autosomal Dominant Arteriopathy with Subcortical Infarcts and \n",
      "Leukoencephalopathy (CADASIL) is the best understood cause of dominantly \n",
      "inherited stroke and results from NOTCH3 mutations that lead to NOTCH3 protein \n",
      "accumulation and selective arterial smooth muscle degeneration. Previous studies \n",
      "show that NOTCH3 protein forms multimers. Here, we investigate protein \n",
      "interactions between NOTCH3 and other vascular Notch isoforms and characterize \n",
      "the effects of elevated NOTCH3 on smo...\n",
      "\n",
      "> Source (Doc id: fd565335-43e3-4574-b636-6912d8b05c57): The gonad arms of C. elegans hermaphrodites acquire invariant shapes by guided \n",
      "migrations of distal tip cells (DTCs), which occur in three phases that differ \n",
      "in the direction and basement membrane substrata used for movement. We found \n",
      "that mig-6 encodes long (MIG-6L) and short (MIG-6S) isoforms of the \n",
      "extracellular matrix protein papilin, each required for distinct aspects of DTC \n",
      "migration. Both MIG-6 isoforms have a predicted N-terminal papilin cassette, \n",
      "lagrin repeats and C-terminal K...\n",
      "\n",
      "> Source (Doc id: aa6598bc-78db-4225-b839-e4a1f4819b8e): Cell invasion through basement membrane is a specialized cellular behavior \n",
      "critical for many developmental processes and leukocyte trafficking. Invasive \n",
      "cellular behavior is also inappropriately co-opted during cancer progression. \n",
      "Acquisition of an invasive phenotype is accompanied by changes in gene \n",
      "expression that are thought to coordinate the steps of invasion. The \n",
      "transcription factors responsible for these changes in gene expression, however, \n",
      "are largely unknown. C. elegans anchor ...\n",
      "\n",
      "> Source (Doc id: cf7e3dd2-92e9-46a6-b901-505f47ed2a6c): The Drosophila dishevelled gene (dsh) encodes a secreted glycoprotein, which \n",
      "regulates cell proliferation, acting as a transducer molecule for developmental \n",
      "processes, including segmentation and neuroblast specification. We have isolated \n",
      "and characterized cDNA clones from two different human dsh-homologous genes, \n",
      "designated as DVL-1 and DVL-3. DVL-1 and DVL-3 putative protein products show \n",
      "64% amino acid identity. The DVL-1 product is 50% identical to dsh and 92% to a \n",
      "murine dsh homolog...\n",
      "\n",
      "> Source (Doc id: 3844e116-6e6d-47c8-95ef-f32f8ea59269): Dishevelled (Dvl) proteins are key transducers of Wnt signaling encoded by \n",
      "members of a multi-gene family in vertebrates. We report here the divergent, \n",
      "tissue-specific expression patterns for all three Dvl genes in Xenopus embryos, \n",
      "which contrast dramatically with their expression patterns in mice. Moreover, we \n",
      "find that the expression patterns of Dvl genes in the chick diverge \n",
      "significantly from those of Xenopus. In addition, in hemichordates, an outgroup \n",
      "to chordates, we find that the...\n"
     ]
    }
   ],
   "source": [
    "response, intermediates = p.run_with_intermediates(query=queries[0]['question'])\n",
    "print(response, response.get_formatted_sources(length = 500))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'query_str': ChatResponse(message=ChatMessage(role=<MessageRole.ASSISTANT: 'assistant'>, content='Can you provide information on whether Hirschsprung disease is classified as a mendelian disorder or a multifactorial disorder?', additional_kwargs={}), raw=ChatCompletion(id='chatcmpl-Aa6er0YvC5meXArUQOYu1j5GAZAYS', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='Can you provide information on whether Hirschsprung disease is classified as a mendelian disorder or a multifactorial disorder?', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None), content_filter_results={'hate': {'filtered': False, 'severity': 'safe'}, 'self_harm': {'filtered': False, 'severity': 'safe'}, 'sexual': {'filtered': False, 'severity': 'safe'}, 'violence': {'filtered': False, 'severity': 'safe'}})], created=1733168765, model='gpt-35-turbo-16k', object='chat.completion', service_tier=None, system_fingerprint=None, usage=CompletionUsage(completion_tokens=25, prompt_tokens=57, total_tokens=82, completion_tokens_details=None, prompt_tokens_details=None), prompt_filter_results=[{'prompt_index': 0, 'content_filter_results': {'hate': {'filtered': False, 'severity': 'safe'}, 'self_harm': {'filtered': False, 'severity': 'safe'}, 'sexual': {'filtered': False, 'severity': 'safe'}, 'violence': {'filtered': False, 'severity': 'safe'}}}]), delta=None, logprobs=None, additional_kwargs={'prompt_tokens': 57, 'completion_tokens': 25, 'total_tokens': 82}),\n",
       " 'nodes': [NodeWithScore(node=TextNode(id_='972a2f43-ddca-4340-ad71-1b5c00ed6e37', embedding=None, metadata={'id': 23001136}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=['id'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='23001136', node_type='4', metadata={'id': 23001136}, hash='d0aaad006146e334dbf6c0132aa1e5cfe1ab55bf32fe33df95aae7abeb5b18e1')}, metadata_template='{key}: {value}', metadata_separator='\\n', text=\"Hirschsprung's disease (HSCR) is a fairly frequent cause of intestinal \\nobstruction in children. It is characterized as a sex-linked heterogonous \\ndisorder with variable severity and incomplete penetrance giving rise to a \\nvariable pattern of inheritance. Although Hirschsprung's disease occurs as an \\nisolated phenotype in at least 70% of cases, it is not infrequently associated \\nwith a number of congenital abnormalities and associated syndromes, \\ndemonstrating a spectrum of congenital anomalies. Certain of these syndromic \\nphenotypes have been linked to distinct genetic sites, indicating underlying \\ngenetic associations of the disease and probable gene-gene interaction, in its \\npathogenesis. These associations with HSCR include Down's syndrome and other \\nchromosomal anomalies, Waardenburg syndrome and other Dominant sensorineural \\ndeafness, the Congenital Central Hypoventilation and Mowat-Wilson and other \\nbrain-related syndromes, as well as the MEN2 and other tumour associations. A \\nnumber of other autosomal recessive syndromes include the Shah-Waardenburg, the \\nBardet-Biedl and Cartilage-hair hypoplasia, Goldberg-Shprintzen syndromes and \\nother syndromes related to cholesterol and fat metabolism among others. The \\ngenetics of Hirschsprung's disease are highly complex with the majority of known \\ngenetic sites relating to the main susceptibility pathways (RET an EDNRB). \\nNon-syndromic non-familial, short-segment HSCR appears to represent a \\nnon-Mendelian condition with variable expression and sex-dependent penetrance. \\nSyndromic and familial forms, on the other hand, have complex patterns of \\ninheritance and being reported as autosomal dominant, recessive and polygenic \\npatterns of inheritance. The phenotypic variability and incomplete penetrance \\nobserved in Hirschsprung's disease could also be explained by the involvement of \\nmodifier genes, especially in its syndromic forms. In this review, we look at \\nthe chromosomal and Mendelian associations and their underlying signalling \\npathways, to obtain a better understanding of the pathogenetic mechanisms \\ninvolved in developing aganglionosis of the distal bowel.\", mimetype='text/plain', start_char_idx=0, end_char_idx=2147, metadata_seperator='\\n', text_template='{metadata_str}\\n\\n{content}'), score=0.99584997),\n",
       "  NodeWithScore(node=TextNode(id_='1ca2b903-27e8-4e4f-bd1c-c49ccc5ec116', embedding=None, metadata={'id': 12239580}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=['id'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='12239580', node_type='4', metadata={'id': 12239580}, hash='0937387ae07f6e1089a704971a76f636ca12f647e92cf90e422842f319ed76a9')}, metadata_template='{key}: {value}', metadata_separator='\\n', text='Hirschsprung disease (HSCR), or congenital intestinal aganglionosis, is a common \\nhereditary disorder causing intestinal obstruction, thereby showing considerable \\nphenotypic variation in conjunction with complex inheritance. Moreover, \\nphenotypic assessment of the disease has been complicated since a subset of the \\nobserved mutations is also associated with several additional syndromic \\nanomalies. Coding sequence mutations in e.g. RET, GDNF, EDNRB, EDN3, and SOX10 \\nlead to long-segment (L-HSCR) as well as syndromic HSCR but fail to explain the \\ntransmission of the much more common short-segment form (S-HSCR). Furthermore, \\nmutations in the RET gene are responsible for approximately half of the familial \\nand some sporadic cases, strongly suggesting, on the one hand, the importance of \\nnon-coding variations and, on the other hand, that additional genes involved in \\nthe development of the enteric nervous system still await their discovery. For \\nalmost all of the identified HSCR genes incomplete penetrance of the HSCR \\nphenotype has been reported, probably due to modifier loci. Therefore, HSCR has \\nbecome a model for a complex oligo-/polygenic disorder in which the relationship \\nbetween different genes creating a non-mendelian inheritance pattern still \\nremains to be elucidated.', mimetype='text/plain', start_char_idx=0, end_char_idx=1296, metadata_seperator='\\n', text_template='{metadata_str}\\n\\n{content}'), score=0.99356186),\n",
       "  NodeWithScore(node=TextNode(id_='110ef15a-726a-44cd-afc7-b8f845dd1c82', embedding=None, metadata={'id': 15617541}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=['id'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='15617541', node_type='4', metadata={'id': 15617541}, hash='234d2e52f84da124e243fd256230e75d511f70e1fe623ca9216947427f02bbe8')}, metadata_template='{key}: {value}', metadata_separator='\\n', text=\"Hirschsprung's disease is characterized by the absence of ganglion cells in the \\nmyenteric and submucosal plexuses of the gastrointestinal tract. Genetic \\ndissection was successful as nine genes and four loci for Hirschsprung's disease \\nsusceptibility were identified. Different approaches were used to find these \\nloci such as classical linkage in large families, identity by descent mapping in \\nan inbred kindred, candidate gene approaches based on naturally occurring mutant \\nmice models, and finally the use of model-free linkage and association analyzes. \\nIn this study, we review the identification of genes and loci involved in the \\nnon-syndromic common form and syndromic Mendelian forms of Hirschsprung's \\ndisease. The majority of the identified genes are related to Mendelian syndromic \\nforms of Hirschsprung's disease. The non-Mendelian inheritance of sporadic \\nnon-syndromic Hirschsprung's disease proved to be complex; involvement of \\nmultiple loci was demonstrated in a multiplicative model. We discuss the \\npractical implications of the elucidation of genes associated with \\nHirschsprung's disease susceptibility for genetic counseling. Finally, we \\nspeculate on possible strategies to identify new genes for Hirschsprung's \\ndisease.\", mimetype='text/plain', start_char_idx=0, end_char_idx=1248, metadata_seperator='\\n', text_template='{metadata_str}\\n\\n{content}'), score=0.9904406),\n",
       "  NodeWithScore(node=TextNode(id_='e0ce075b-b922-4b28-b1ab-853b0f13f58f', embedding=None, metadata={'id': 20598273}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=['id'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='20598273', node_type='4', metadata={'id': 20598273}, hash='67423c64b90a258a2f8972d537963cb89ad1ea3feefec1a3cb143ab9f20f3397')}, metadata_template='{key}: {value}', metadata_separator='\\n', text='The major gene for Hirschsprung disease (HSCR) encodes the receptor tyrosine \\nkinase RET. In a study of 690 European- and 192 Chinese-descent probands and \\ntheir parents or controls, we demonstrate the ubiquity of a >4-fold \\nsusceptibility from a C-->T allele (rs2435357: p = 3.9 x 10(-43) in European \\nancestry; p = 1.1 x 10(-21) in Chinese samples) that probably arose once within \\nthe intronic RET enhancer MCS+9.7. With in vitro assays, we now show that the T \\nvariant disrupts a SOX10 binding site within MCS+9.7 that compromises RET \\ntransactivation. The T allele, with a control frequency of 20%-30%/47% and case \\nfrequency of 54%-62%/88% in European/Chinese-ancestry individuals, is involved \\nin all forms of HSCR. It is marginally associated with proband gender (p = 0.13) \\nand significantly so with length of aganglionosis (p = 7.6 x 10(-5)) and \\nfamiliality (p = 6.2 x 10(-4)). The enhancer variant is more frequent in the \\ncommon forms of male, short-segment, and simplex families whereas multiple, \\nrare, coding mutations are the norm in the less common and more severe forms of \\nfemale, long-segment, and multiplex families. The T variant also increases \\npenetrance in patients with rare RET coding mutations. Thus, both rare and \\ncommon mutations, individually and together, make contributions to the risk of \\nHSCR. The distribution of RET variants in diverse HSCR patients suggests a \\n\"cellular-recessive\" genetic model where both RET alleles\\' function is \\ncompromised. The RET allelic series, and its genotype-phenotype correlations, \\nshows that success in variant identification in complex disorders may strongly \\ndepend on which patients are studied.', mimetype='text/plain', start_char_idx=0, end_char_idx=1669, metadata_seperator='\\n', text_template='{metadata_str}\\n\\n{content}'), score=0.95589906),\n",
       "  NodeWithScore(node=TextNode(id_='2360eafe-1e46-4cc1-8771-2ae4ce088434', embedding=None, metadata={'id': 24040443}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=['id'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='24040443', node_type='4', metadata={'id': 24040443}, hash='85b4da9847c652ca6b00aba4fe9eba11776513925bba1597e5b20f59104a2b95')}, metadata_template='{key}: {value}', metadata_separator='\\n', text=\"Hirschsprung's disease (HSCR) is a congenital disorder of the enteric nervous \\nsystem and is characterized by an absence of enteric ganglion cells in terminal \\nregions of the gut during development. Dishevelled (DVL) protein is a \\ncytoplasmic protein which plays pivotal roles in the embryonic development. In \\nthis study, we explore the cause of HSCR by studying the expression of DVL-1 and \\nDVL-3 genes and their proteins in the aganglionic segment and the ganglionic \\nsegment of colon in HSCR patients.\\nMATERIALS AND METHODS: Specimen of aganglionic segment and ganglionic segment of \\ncolon in 50 cases of HSCR patients. Expression levels of mRNA and proteins of \\nDVL-1 and DVL-3 were confirmed by quantitative real-time PCR (qRT-PCR), western \\nblot and immunohistochemistry staining between the aganglionic segment and the \\nganglionic segment of colon in HSCR patients.\\nRESULTS: The mRNA expression of DVL-1 and DVL-3 were 2.06 fold and 3.12 fold in \\nthe aganglionic segment colon tissues compared to the ganglionic segment, \\nrespectively. Similarly, the proteins expression of DVL-1 and DVL-3 were higher \\n(39.71 ± 4.53 vs and 53.90 ± 6.79 vs) in the aganglionic segment colon tissues \\nthan in the ganglionic segment (15.01 ± 2.66 and 20.13 ± 3.63) by western blot. \\nBesides, immunohistochemical staining showed that DVL-1 and DVL-3 have a \\nsignificant increase in mucous and submucous layers from aganglionic colon \\nsegments compared with ganglionic segments.\\nCONCLUSION: The study showed an association of DVL-1 and DVL-3 with HSCR, it may \\nplay an important role in the pathogenesis of HSCR.\", mimetype='text/plain', start_char_idx=0, end_char_idx=1599, metadata_seperator='\\n', text_template='{metadata_str}\\n\\n{content}'), score=0.81405276),\n",
       "  NodeWithScore(node=TextNode(id_='7aa70398-da0c-460b-ab42-061f486743e8', embedding=None, metadata={'id': 23028706}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=['id'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='23028706', node_type='4', metadata={'id': 23028706}, hash='b6de9f24211c16e19eed91bb9e29ad4089b5593df1adffa7d5a0c6748b5bb913')}, metadata_template='{key}: {value}', metadata_separator='\\n', text='Cerebral Autosomal Dominant Arteriopathy with Subcortical Infarcts and \\nLeukoencephalopathy (CADASIL) is the best understood cause of dominantly \\ninherited stroke and results from NOTCH3 mutations that lead to NOTCH3 protein \\naccumulation and selective arterial smooth muscle degeneration. Previous studies \\nshow that NOTCH3 protein forms multimers. Here, we investigate protein \\ninteractions between NOTCH3 and other vascular Notch isoforms and characterize \\nthe effects of elevated NOTCH3 on smooth muscle gene regulation. We demonstrate \\nthat NOTCH3 forms heterodimers with NOTCH1, NOTCH3, and NOTCH4. R90C and C49Y \\nmutant NOTCH3 form complexes which are more resistant to detergents than wild \\ntype NOTCH3 complexes. Using quantitative NOTCH3-luciferase clearance assays, we \\nfound significant inhibition of mutant NOTCH3 clearance. In coculture assays of \\nNOTCH function, overexpressed wild type and mutant NOTCH3 significantly \\nrepressed NOTCH-regulated smooth muscle transcripts and potently impaired the \\nactivity of three independent smooth muscle promoters. Wildtype and R90C \\nrecombinant NOTCH3 proteins applied to cell cultures also blocked canonical \\nNotch fuction. We conclude that CADASIL mutants of NOTCH3 complex with NOTCH1, \\n3, and 4, slow NOTCH3 clearance, and that overexpressed wild type and mutant \\nNOTCH3 protein interfere with key NOTCH-mediated functions in smooth muscle \\ncells.', mimetype='text/plain', start_char_idx=0, end_char_idx=1406, metadata_seperator='\\n', text_template='{metadata_str}\\n\\n{content}'), score=0.16844618),\n",
       "  NodeWithScore(node=TextNode(id_='fd565335-43e3-4574-b636-6912d8b05c57', embedding=None, metadata={'id': 19297413}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=['id'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='19297413', node_type='4', metadata={'id': 19297413}, hash='c8fa710612ae83a6c83a36de91e6926767a6e79fd8c3df29f33446c2eec3d6c2')}, metadata_template='{key}: {value}', metadata_separator='\\n', text='The gonad arms of C. elegans hermaphrodites acquire invariant shapes by guided \\nmigrations of distal tip cells (DTCs), which occur in three phases that differ \\nin the direction and basement membrane substrata used for movement. We found \\nthat mig-6 encodes long (MIG-6L) and short (MIG-6S) isoforms of the \\nextracellular matrix protein papilin, each required for distinct aspects of DTC \\nmigration. Both MIG-6 isoforms have a predicted N-terminal papilin cassette, \\nlagrin repeats and C-terminal Kunitz-type serine proteinase inhibitory domains. \\nWe show that mutations affecting MIG-6L specifically and cell-autonomously \\ndecrease the rate of post-embryonic DTC migration, mimicking a post-embryonic \\ncollagen IV deficit. We also show that MIG-6S has two separable functions - one \\nin embryogenesis and one in the second phase of DTC migration. Genetic data \\nsuggest that MIG-6S functions in the same pathway as the MIG-17/ADAMTS \\nmetalloproteinase for guiding phase 2 DTC migrations, and MIG-17 is abnormally \\nlocalized in mig-6 class-s mutants. Genetic data also suggest that MIG-6S and \\nnon-fibrillar network collagen IV play antagonistic roles to ensure normal phase \\n2 DTC guidance.', mimetype='text/plain', start_char_idx=0, end_char_idx=1188, metadata_seperator='\\n', text_template='{metadata_str}\\n\\n{content}'), score=0.07250525),\n",
       "  NodeWithScore(node=TextNode(id_='aa6598bc-78db-4225-b839-e4a1f4819b8e', embedding=None, metadata={'id': 21784067}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=['id'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='21784067', node_type='4', metadata={'id': 21784067}, hash='e0d291773c7d325930192692fe271b5de7cb691029d5000eb45eb8b9ab746fbd')}, metadata_template='{key}: {value}', metadata_separator='\\n', text='Cell invasion through basement membrane is a specialized cellular behavior \\ncritical for many developmental processes and leukocyte trafficking. Invasive \\ncellular behavior is also inappropriately co-opted during cancer progression. \\nAcquisition of an invasive phenotype is accompanied by changes in gene \\nexpression that are thought to coordinate the steps of invasion. The \\ntranscription factors responsible for these changes in gene expression, however, \\nare largely unknown. C. elegans anchor cell (AC) invasion is a genetically \\ntractable in vivo model of invasion through basement membrane. AC invasion \\nrequires the conserved transcription factor FOS-1A, but other transcription \\nfactors are thought to act in parallel to FOS-1A to control invasion. Here we \\nidentify the transcription factor HLH-2, the C. elegans ortholog of Drosophila \\nDaughterless and vertebrate E proteins, as a regulator of AC invasion. Reduction \\nof HLH-2 function by RNAi or with a hypomorphic allele causes defects in AC \\ninvasion. Genetic analysis indicates that HLH-2 has functions outside of the \\nFOS-1A pathway. Using expression analysis, we identify three genes that are \\ntranscriptionally regulated by HLH-2: the protocadherin cdh-3, and two genes \\nencoding secreted extracellular matrix proteins, mig-6/papilin and \\nhim-4/hemicentin. Further, we show that reduction of HLH-2 function causes \\ndefects in polarization of F-actin to the invasive cell membrane, a process \\nrequired for the AC to generate protrusions that breach the basement membrane. \\nThis work identifies HLH-2 as a regulator of the invasive phenotype in the AC, \\nadding to our understanding of the transcriptional networks that control cell \\ninvasion.', mimetype='text/plain', start_char_idx=0, end_char_idx=1707, metadata_seperator='\\n', text_template='{metadata_str}\\n\\n{content}'), score=0.053899154),\n",
       "  NodeWithScore(node=TextNode(id_='cf7e3dd2-92e9-46a6-b901-505f47ed2a6c', embedding=None, metadata={'id': 8817329}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=['id'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='8817329', node_type='4', metadata={'id': 8817329}, hash='a0ff61e2367024cb589e3c25196f26114be719de964c8bf8ccd8dafdc7cadeed')}, metadata_template='{key}: {value}', metadata_separator='\\n', text='The Drosophila dishevelled gene (dsh) encodes a secreted glycoprotein, which \\nregulates cell proliferation, acting as a transducer molecule for developmental \\nprocesses, including segmentation and neuroblast specification. We have isolated \\nand characterized cDNA clones from two different human dsh-homologous genes, \\ndesignated as DVL-1 and DVL-3. DVL-1 and DVL-3 putative protein products show \\n64% amino acid identity. The DVL-1 product is 50% identical to dsh and 92% to a \\nmurine dsh homologue (Dvl-1). Both human DVL genes are widely expressed in fetal \\nand adult tissues, including brain, lung, kidney, skeletal muscle and heart. \\nDVL-1 locus maps to chromosome 1p36 and DVL-3 to chromosome 3q27. DVL-1 locus on \\nchromosome 1 corresponds to the murine syntenic region where Dvl-1 is located. \\nDVL-1 and DVL-3 are members of a human dsh-like gene family, which is probably \\ninvolved in human development. Although the precise role of these genes in \\nembryogenesis is only conjectural at present, the structural and evolutionary \\ncharacteristics suggest that mutations at their loci may be involved in neural \\nand heart developmental defects.', mimetype='text/plain', start_char_idx=0, end_char_idx=1148, metadata_seperator='\\n', text_template='{metadata_str}\\n\\n{content}'), score=0.022201229),\n",
       "  NodeWithScore(node=TextNode(id_='3844e116-6e6d-47c8-95ef-f32f8ea59269', embedding=None, metadata={'id': 19618470}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=['id'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='19618470', node_type='4', metadata={'id': 19618470}, hash='eb5ad1e05c810a5abbd036460b2a60984bf66aebf7dde1ce992929c8e6701f0f')}, metadata_template='{key}: {value}', metadata_separator='\\n', text='Dishevelled (Dvl) proteins are key transducers of Wnt signaling encoded by \\nmembers of a multi-gene family in vertebrates. We report here the divergent, \\ntissue-specific expression patterns for all three Dvl genes in Xenopus embryos, \\nwhich contrast dramatically with their expression patterns in mice. Moreover, we \\nfind that the expression patterns of Dvl genes in the chick diverge \\nsignificantly from those of Xenopus. In addition, in hemichordates, an outgroup \\nto chordates, we find that the one Dvl gene is dynamically expressed in a \\ntissue-specific manner. Using knockdowns, we find that Dvl1 and Dvl2 are \\nrequired for early neural crest specification and for somite segmentation in \\nXenopus. Most strikingly, we report a novel role for Dvl3 in the maintenance of \\ngene expression in muscle and in the development of the Xenopus sclerotome. \\nThese data demonstrate that the expression patterns and developmental functions \\nof specific Dvl genes have diverged significantly during chordate evolution.', mimetype='text/plain', start_char_idx=0, end_char_idx=1009, metadata_seperator='\\n', text_template='{metadata_str}\\n\\n{content}'), score=0.022074386)]}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intermediates['summarizer'].inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>ap</th>\n",
       "      <th>ndcg</th>\n",
       "      <th>hit_rate</th>\n",
       "      <th>mrr</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k</th>\n",
       "      <th>splitter</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">5</th>\n",
       "      <th>sentance_512_0</th>\n",
       "      <td>0.920000</td>\n",
       "      <td>0.484722</td>\n",
       "      <td>0.484722</td>\n",
       "      <td>0.947518</td>\n",
       "      <td>0.484722</td>\n",
       "      <td>0.482333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>token_512</th>\n",
       "      <td>0.920000</td>\n",
       "      <td>0.464722</td>\n",
       "      <td>0.484722</td>\n",
       "      <td>0.947518</td>\n",
       "      <td>0.484722</td>\n",
       "      <td>0.482333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">10</th>\n",
       "      <th>sentance_512_0</th>\n",
       "      <td>0.593333</td>\n",
       "      <td>0.567222</td>\n",
       "      <td>0.587222</td>\n",
       "      <td>0.715440</td>\n",
       "      <td>0.587222</td>\n",
       "      <td>0.432328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>token_512</th>\n",
       "      <td>0.622222</td>\n",
       "      <td>0.567222</td>\n",
       "      <td>0.607222</td>\n",
       "      <td>0.731120</td>\n",
       "      <td>0.607222</td>\n",
       "      <td>0.422661</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   precision    recall        ap      ndcg  hit_rate       mrr\n",
       "k  splitter                                                                   \n",
       "5  sentance_512_0   0.920000  0.484722  0.484722  0.947518  0.484722  0.482333\n",
       "   token_512        0.920000  0.464722  0.484722  0.947518  0.484722  0.482333\n",
       "10 sentance_512_0   0.593333  0.567222  0.587222  0.715440  0.587222  0.432328\n",
       "   token_512        0.622222  0.567222  0.607222  0.731120  0.607222  0.422661"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from llama_index.core.evaluation import RetrieverEvaluator\n",
    "from llama_index.core.evaluation.retrieval.metrics import resolve_metrics, HitRate, MRR\n",
    "from llama_index.core.node_parser import SentenceSplitter, TokenTextSplitter\n",
    "\n",
    "metric_dict = {}\n",
    "metrics = [\"precision\", \"recall\", \"ap\", \"ndcg\"]\n",
    "metrics = [x() for x in resolve_metrics(metrics)] + [HitRate(use_granular_hit_rate=True), MRR(use_granular_mrr=True)]\n",
    "\n",
    "results_data = []\n",
    "splitters = {'sentance_512_0': SentenceSplitter(chunk_size=512, chunk_overlap=0),\n",
    "             'token_512': TokenTextSplitter(chunk_size=512, chunk_overlap=100)}\n",
    "\n",
    "for splitter_name in splitters.keys():\n",
    "    splitter = splitters[splitter_name]\n",
    "    index = VectorStoreIndex.from_documents(docs, transformations = [splitter])\n",
    "    \n",
    "    for k in [5, 10]:\n",
    "        p = prepare_pipeline(new_index = index, k=k, only_retriever=True) # we only need to eval QR+RR retriever in this loop\n",
    "        p.verbose = False\n",
    "        for row in queries:\n",
    "            row['relevant_passage_ids'] = row['relevant_passage_ids'][1:-1].split(', ')\n",
    "            query = row['question']\n",
    "            retrieved_nodes = p.run(query) # without llm it will return the nodes from reranker \n",
    "            retrieved_passage_ids = [str(node.metadata['id']) for node in retrieved_nodes]\n",
    "            \n",
    "            for metric in metrics:\n",
    "                eval_result = metric.compute(\n",
    "                    query, row['relevant_passage_ids'], retrieved_passage_ids,\n",
    "                )\n",
    "                metric_dict[metric.metric_name] = eval_result.score\n",
    "            \n",
    "            results_data.append({\n",
    "                'splitter': splitter_name,\n",
    "                'k': k,\n",
    "                'query': query,\n",
    "                'retrieved_ids': retrieved_passage_ids,\n",
    "                'relevant_ids': row['relevant_passage_ids'],\n",
    "                **metric_dict\n",
    "            })\n",
    "\n",
    "results_df = pd.DataFrame(results_data)\n",
    "results_df.drop(['query', 'retrieved_ids', 'relevant_ids'], axis=1).groupby(['k','splitter']).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Whiplash\\AppData\\Local\\Temp\\ipykernel_13952\\41827363.py:17: LangChainDeprecationWarning: The class `AzureChatOpenAI` was deprecated in LangChain 0.0.10 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-openai package and should be used instead. To use it run `pip install -U :class:`~langchain-openai` and import as `from :class:`~langchain_openai import AzureChatOpenAI``.\n",
      "  evaluator_llm = LangchainLLMWrapper(AzureChatOpenAI(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "required_keys={'query_str', 'nodes'} optional_keys=set()\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "decabbb025d34063bcd9b924f40306d7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/30 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "required_keys={'query_str', 'nodes'} optional_keys=set()\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bc2b0aec240645b08011f6f89f8bb482",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/30 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "required_keys={'query_str', 'nodes'} optional_keys=set()\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6037f6aa3b544451917e9d8c0858c428",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/30 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "required_keys={'query_str', 'nodes'} optional_keys=set()\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "013c6b66761140b88c0c01e13b7427af",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/30 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>context_recall</th>\n",
       "      <th>llm_context_precision_with_reference</th>\n",
       "      <th>factual_correctness</th>\n",
       "      <th>semantic_similarity</th>\n",
       "      <th>answer_correctness</th>\n",
       "      <th>faithfulness</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k</th>\n",
       "      <th>splitter</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">5</th>\n",
       "      <th>sentance_512_0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.980000</td>\n",
       "      <td>0.634</td>\n",
       "      <td>0.919111</td>\n",
       "      <td>0.494778</td>\n",
       "      <td>0.571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>token_512</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.980000</td>\n",
       "      <td>0.706</td>\n",
       "      <td>0.917222</td>\n",
       "      <td>0.510972</td>\n",
       "      <td>0.571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">10</th>\n",
       "      <th>sentance_512_0</th>\n",
       "      <td>0.9</td>\n",
       "      <td>0.918255</td>\n",
       "      <td>0.656</td>\n",
       "      <td>0.916280</td>\n",
       "      <td>0.511913</td>\n",
       "      <td>0.483333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>token_512</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.913872</td>\n",
       "      <td>0.694</td>\n",
       "      <td>0.919359</td>\n",
       "      <td>0.476268</td>\n",
       "      <td>0.571429</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   context_recall  llm_context_precision_with_reference  \\\n",
       "k  splitter                                                               \n",
       "5  sentance_512_0             1.0                              0.980000   \n",
       "   token_512                  1.0                              0.980000   \n",
       "10 sentance_512_0             0.9                              0.918255   \n",
       "   token_512                  1.0                              0.913872   \n",
       "\n",
       "                   factual_correctness  semantic_similarity  \\\n",
       "k  splitter                                                   \n",
       "5  sentance_512_0                0.634             0.919111   \n",
       "   token_512                     0.706             0.917222   \n",
       "10 sentance_512_0                0.656             0.916280   \n",
       "   token_512                     0.694             0.919359   \n",
       "\n",
       "                   answer_correctness  faithfulness  \n",
       "k  splitter                                          \n",
       "5  sentance_512_0            0.494778      0.571429  \n",
       "   token_512                 0.510972      0.571429  \n",
       "10 sentance_512_0            0.511913      0.483333  \n",
       "   token_512                 0.476268      0.571429  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from ragas.dataset_schema import SingleTurnSample\n",
    "                            \n",
    "from ragas.metrics import (LLMContextRecall,LLMContextPrecisionWithReference, Faithfulness, \n",
    "                            SemanticSimilarity, NonLLMContextRecall, answer_correctness, FactualCorrectness)\n",
    "from ragas import evaluate, EvaluationDataset\n",
    "from langchain_openai import AzureOpenAIEmbeddings\n",
    "from langchain.chat_models import AzureChatOpenAI\n",
    "from ragas.llms import LangchainLLMWrapper\n",
    "from ragas.embeddings import LangchainEmbeddingsWrapper\n",
    "\n",
    "# evaluator_llm = LangchainLLMWrapper(AzureChatOpenAI(\n",
    "#                 openai_api_version=os.environ['AZURE_OPENAI_API_VERSION'],\n",
    "#                 azure_deployment='gpt-4o',\n",
    "#                 model='gpt-4o',\n",
    "#             ))\n",
    "\n",
    "evaluator_llm = LangchainLLMWrapper(AzureChatOpenAI(\n",
    "                openai_api_version=os.environ['AZURE_OPENAI_API_VERSION'],\n",
    "                azure_deployment='gpt-35-turbo16k',\n",
    "                model='gpt-35-turbo',\n",
    "            ))\n",
    "\n",
    "evaluator_embeddings = LangchainEmbeddingsWrapper( AzureOpenAIEmbeddings(\n",
    "                openai_api_version=os.environ['AZURE_OPENAI_API_VERSION'],\n",
    "                azure_deployment='text-embedding-ada-002',\n",
    "                model='text-embedding-ada-002',\n",
    "))\n",
    "\n",
    "\n",
    "metrics = [\n",
    "    LLMContextRecall(), # Recall based on claims made in response vs those in reference, uses LLM\n",
    "    LLMContextPrecisionWithReference(), # Precision based on claims made in response vs those in reference, uses LLM\n",
    "    FactualCorrectness(), # F1-Score of claims made in response vs those in reference\n",
    "    SemanticSimilarity(), # embedding based similarity between generated answer and ground truth\n",
    "    answer_correctness,\n",
    "    Faithfulness()\n",
    "]\n",
    "\n",
    "results_data = []\n",
    "splitters = {'sentance_512_0': SentenceSplitter(chunk_size=512, chunk_overlap=0),\n",
    "             'token_512': TokenTextSplitter(chunk_size=512, chunk_overlap=100)}\n",
    "\n",
    "for splitter_name in splitters.keys():\n",
    "    splitter = splitters[splitter_name]\n",
    "    index = VectorStoreIndex.from_documents(docs, transformations = [splitter])\n",
    "    \n",
    "    for k in [5, 10]:\n",
    "        p = prepare_pipeline(new_index = index, k=k) # we only need to eval QR+RR retriever in this loop\n",
    "        p.verbose = False\n",
    "        samples = []\n",
    "        for row in queries:\n",
    "            query = row['question']\n",
    "            response = p.run(query)\n",
    "            retrieved_nodes = response.source_nodes\n",
    "            retrieved_passage_ids = [node.metadata['id'] for node in retrieved_nodes]\n",
    "            retrieved_passages = [ ds.loc[int(id)].passage for id in retrieved_passage_ids ]\n",
    "            relevant_passages = [ ds.loc[int(id)].passage for id in row[\"relevant_passage_ids\"][1:-1].split(', ') ]\n",
    "            \n",
    "            sample = SingleTurnSample(\n",
    "                user_input=query,\n",
    "                reference=row[\"answer\"],\n",
    "                response=response.response,\n",
    "                retrieved_contexts=retrieved_passages,\n",
    "                reference_contexts=relevant_passages,\n",
    "            )\n",
    "            samples.append(sample)\n",
    "\n",
    "        eval_dataset = EvaluationDataset(samples = samples)\n",
    "        results = evaluate(dataset=eval_dataset, metrics=metrics, llm = evaluator_llm, embeddings = evaluator_embeddings)\n",
    "        df = results.to_pandas()\n",
    "        df['k'] = k\n",
    "        df['splitter'] = splitter_name\n",
    "        results_data.append(df)\n",
    "\n",
    "results_df = pd.concat(results_data).reset_index(drop=True)\n",
    "results_df.drop(['user_input','retrieved_contexts','reference_contexts','response','reference'], axis=1).groupby(['k','splitter']).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dl",
   "language": "python",
   "name": "dl"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
